src/uvarint.cr

```diff
@@ -1,44 +1,131 @@
+require "big/big_int"
+
 require "./uvarint/*"
 
 # This module is based on the varint implementation in the Go programming
 # language. See: https://golang.org/src/encoding/binary/varint.go
 
-MSB = 0x80_u8  # 10000000
-REST = 0x7F_u8 # 01111111
-
-module UVarint
-  extend self
+private MSB = 0x80_u8
+private REST = 0x7F_u8
+private ZERO = 0.to_big_i
 
-  alias UVarInt = Array(UInt8)
+# Encodes an Int::Unsined into a Bytes slice.
+private def encode(n : Int::Unsigned | BigInt) : Bytes
+  arr = [] of UInt8
+  while n >= MSB
+    arr << (n.to_u8 | MSB)
+    n = n >> 7
+  end
+  arr << n.to_u8
+  ptr = arr.to_unsafe
+  Bytes.new(ptr, arr.size)
+end
 
-  # encode encodes a UInt64 into a UVarInt,
-  # which is an Array of UInt8's.
-  def encode(n : UInt64) : UVarInt
-    buf = [] of UInt8
-    while n >= MSB
-      buf << (n.to_u8 | MSB)
-      n = n >> 7
+# Decodes enumerable bytes into a BigInt.
+# If the enumeration is longer thar 10 bytes,
+# then an overflow exception is raised.
+private def decode(bytes : Bytes) : BigInt
+  x = ZERO
+  s = 0
+  bytes.each_with_index do |b, i|
+    if b < MSB
+      return x | (b.to_big_i << s)
     end
-    buf << n.to_u8
-    return buf
-  end
-
-  # decode decodes a UVarInt into a UInt64.
-  # If the UVarInt contains more than 10 UInt8's,
-  # then an overflow exception is raised.
-  def decode(vi : UVarInt) : UInt64
-    x = 0_u64
-    s = 0
-    vi.each_with_index do |v, i|
-      if v < MSB
-        if i > 9 || i == 9 && v > 1
-          raise Exception.new("overflow")
-        end
-        return x | (v.to_u64 << s)
-      end
-      x |= (v & REST).to_u64 << s
-      s += 7
-    end
-    return 0_u64
+    x = x | ((b & REST).to_big_i << s)
+    s += 7
```

Comment:
@freitagbr the magic isn't so obscure. I'm pretty sure we use 7 because there's 8 bits in a byte. If you shiftright by 7, you get MSB. If you shift left by 7 again, you get MSB followed by all zeros. I haven't reviewed the algorithm in a bit, but I imagine that's the use case.